File: dotnet/Base/GameX.Base/Meta.cs
using GameX.Formats;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameX.Meta
{
    #region MetaContent

    [DebuggerDisplay("{Type}: {Name}")]
    public class MetaContent
    {
        public string Type { get; set; }
        public string Name { get; set; }
        public object Value { get; set; }
        public object Tag { get; set; }
        public int MaxWidth { get; set; }
        public int MaxHeight { get; set; }
        public IDisposable Dispose { get; set; }
        public Type EngineType { get; set; }
    }

    #endregion

    #region MetaInfo

    [DebuggerDisplay("{Name}, items: {Items.Count} [{Tag}]")]
    public class MetaInfo
    {
        public string Name { get; set; }
        public object Tag { get; }
        public IEnumerable<MetaInfo> Items { get; }
        public bool Clickable { get; set; }

        public MetaInfo(string name, object tag = null, IEnumerable<MetaInfo> items = null, bool clickable = false)
        {
            Name = name;
            Tag = tag;
            Items = items ?? Enumerable.Empty<MetaInfo>();
            Clickable = clickable;
        }

        public static MetaInfo WrapWithGroup<T>(IList<T> source, string groupName, IEnumerable<MetaInfo> body)
            => source.Count == 0 ? null
            : source.Count == 1 ? body.First()
            : new MetaInfo(groupName, body);
    }

    #endregion

    #region MetaItem

    [DebuggerDisplay("{Name}, items: {Items.Count}")]
    public class MetaItem
    {
        [DebuggerDisplay("{Name}")]
        public class Filter
        {
            public string Name;
            public string Description;

            public Filter(string name, string description = "")
            {
                Name = name;
                Description = description;
            }

            public override string ToString() => Name;
        }

        public object Source { get; }
        public string Name { get; }
        public object Icon { get; }
        public object Tag { get; }
        public PakFile PakFile { get; }
        public List<MetaItem> Items { get; private set; }

        public MetaItem(object source, string name, object icon, object tag = null, PakFile pakFile = null, List<MetaItem> items = null)
        {
            Source = source;
            Name = name;
            Icon = icon;
            Tag = tag;
            PakFile = pakFile;
            Items = items ?? new List<MetaItem>();
        }

        public MetaItem Search(Func<MetaItem, bool> predicate)
        {
            // if node is a leaf
            if (Items == null || Items.Count == 0) return predicate(this) ? this : null;
            // Otherwise if node is not a leaf
            else
            {
                var results = Items.Select(i => i.Search(predicate)).Where(i => i != null).ToList();
                if (results.Any())
                {
                    var result = (MetaItem)MemberwiseClone();
                    result.Items = results;
                    return result;
                }
                return null;
            }
        }

        public MetaItem FindByPath(string path, MetaManager manager)
        {
            var paths = path.Split(new[] { '\\', '/', ':' }, 2);
            var node = Items.FirstOrDefault(x => x.Name == paths[0]);
            (node?.Source as FileSource)?.Pak?.Open(node.Items, manager);
            //node?.PakFile?.Open(node.Items, manager);
            return node == null || paths.Length == 1 ? node : node.FindByPath(paths[1], manager);
        }

        public static MetaItem FindByPathForNodes(List<MetaItem> nodes, string path, MetaManager manager)
        {
            var paths = path.Split(new[] { '\\', '/', ':' }, 2);
            var node = nodes.FirstOrDefault(x => x.Name == paths[0]);
            (node?.Source as FileSource)?.Pak?.Open(node.Items, manager);
            //node?.PakFile?.Open(node.Items, manager);
            return node == null || paths.Length == 1 ? node : node.FindByPath(paths[1], manager);
        }
    }

    #endregion

    #region IHaveMetaInfo

    public interface IHaveMetaInfo
    {
        List<MetaInfo> GetInfoNodes(MetaManager resource = null, FileSource file = null, object tag = null);
    }

    #endregion

    #region MetaManager

    public abstract class MetaManager
    {
        public abstract object FolderIcon { get; }
        public abstract object PackageIcon { get; }
        public abstract object GetIcon(string name);
        public abstract object GetImage(string name);

        /// <summary>
        /// Gets the string or bytes.
        /// </summary>
        /// <param name="stream">The stream.</param>
        /// <param name="stream">The stream.</param>
        /// <param name="dispose">The dispose.</param>
        /// <returns></returns>
        public static object GuessStringOrBytes(Stream stream, bool dispose = true)
        {
            using var ms = new MemoryStream();
            stream.Position = 0;
            stream.CopyTo(ms);
            var bytes = ms.ToArray();
            if (dispose) stream.Dispose();
            return !bytes.Contains<byte>(0x00)
                ? (object)Encoding.UTF8.GetString(bytes)
                : bytes;
        }

        /// <summary>
        /// Gets the explorer information nodes.
        /// </summary>
        /// <param name="manager">The manager.</param>
        /// <param name="pakFile">The pak file.</param>
        /// <param name="file">The file.</param>
        /// <returns></returns>
        /// <exception cref="NotImplementedException"></exception>
        public static async Task<List<MetaInfo>> GetMetaInfos(MetaManager manager, BinaryPakFile pakFile, FileSource file)
        {
            List<MetaInfo> nodes = null;
            var obj = await pakFile.LoadFileObject<object>(file);
            if (obj == null) return null;
            else if (obj is IHaveMetaInfo info) nodes = info.GetInfoNodes(manager, file);
            else if (obj is Stream stream)
            {
                var value = GuessStringOrBytes(stream);
                nodes = value is string text ? new List<MetaInfo> {
                        new MetaInfo(null, new MetaContent { Type = "Text", Name = "Text", Value = text }),
                        new MetaInfo("Text", items: new List<MetaInfo> {
                            new MetaInfo($"Length: {text.Length}"),
                        }) }
                    : value is byte[] bytes ? new List<MetaInfo> {
                        new MetaInfo(null, new MetaContent { Type = "Hex", Name = "Hex", Value = new MemoryStream(bytes) }),
                        new MetaInfo("Bytes", items: new List<MetaInfo> {
                            new MetaInfo($"Length: {bytes.Length}"),
                        }) }
                    : throw new ArgumentOutOfRangeException(nameof(value), value.GetType().Name);
            }
            else if (obj is IDisposable disposable) disposable.Dispose();
            if (nodes == null) return null;
            nodes.Add(new MetaInfo("File", items: new List<MetaInfo> {
                new MetaInfo($"Path: {file.Path}"),
                new MetaInfo($"FileSize: {file.FileSize}"),
                file.Parts != null
                    ? new MetaInfo("Parts", items: file.Parts.Select(part => new MetaInfo($"{part.FileSize}@{part.Path}")))
                    : null
            }));
            //nodes.Add(new MetaInfo(null, new MetaContent { Type = "Hex", Name = "TEST", Value = new MemoryStream() }));
            //nodes.Add(new MetaInfo(null, new MetaContent { Type = "Image", Name = "TEST", MaxWidth = 500, MaxHeight = 500, Value = null }));
            return nodes;
        }

        /// <summary>
        /// Gets the meta items.
        /// </summary>
        /// <param name="manager">The manager.</param>
        /// <param name="pakFile">The pak file.</param>
        /// <returns></returns>
        public static List<MetaItem> GetMetaItems(MetaManager manager, BinaryPakFile pakFile)
        {
            var root = new List<MetaItem>();
            if (pakFile.Files == null || pakFile.Files.Count == 0) return root;
            string currentPath = null; List<MetaItem> currentFolder = null;

            // parse paths
            foreach (var file in pakFile.Files.OrderBy(x => x.Path))
            {
                // next path, skip empty
                var path = file.Path[pakFile.PathSkip..];
                if (string.IsNullOrEmpty(path)) continue;
                // folder
                var fileFolder = Path.GetDirectoryName(path);
                if (currentPath != fileFolder)
                {
                    currentPath = fileFolder;
                    currentFolder = root;
                    if (!string.IsNullOrEmpty(fileFolder))
                        foreach (var folder in fileFolder.Split('\\'))
                        {
                            var found = currentFolder.Find(x => x.Name == folder && x.PakFile == null);
                            if (found != null) currentFolder = found.Items;
                            else
                            {
                                found = new MetaItem(file, folder, manager.FolderIcon);
                                currentFolder.Add(found);
                                currentFolder = found.Items;
                            }
                        }
                }
                // pakfile
                if (file.Pak != null)
                {
                    var items = GetMetaItems(manager, file.Pak);
                    currentFolder.Add(new MetaItem(file, Path.GetFileName(file.Path), manager.PackageIcon, pakFile: pakFile, items: items));
                    continue;
                }
                // file
                var fileName = Path.GetFileName(path);
                var fileNameForIcon = pakFile.FileMask?.Invoke(fileName) ?? fileName;
                var extentionForIcon = Path.GetExtension(fileNameForIcon);
                if (extentionForIcon.Length > 0) extentionForIcon = extentionForIcon[1..];
                currentFolder.Add(new MetaItem(file, fileName, manager.GetIcon(extentionForIcon), pakFile: pakFile));
            }
            return root;
        }
    }

    #endregion
}

File: Python/gamex/meta.py
import os, re, pathlib
from io import BytesIO
from gamex.filesrc import FileSource
from gamex.util import _throw

# typedefs
class PakFile: pass
class BinaryPakFile: pass

# forwards
class MetaInfo: pass
class MetaItem: pass
class MetaManager: pass

# MetaContent
class MetaContent:
    def __init__(self, type: str, name: str, value: object = None, 
        tag: object = None, maxWidth: int = None, maxHeight: int = None,
        dispose: object = None, engineType: type = None):
        self.type = type
        self.name = name
        self.value = value
        self.tag = tag
        self.maxWidth = maxWidth
        self.maxHeight = maxHeight
        self.dispose = dispose
        self.engineType = engineType

# MetaInfo
class MetaInfo:
    def __init__(self, name: str, tag: object = None, items: list[MetaInfo] = None, clickable: bool = False):
        self.name = name
        self.tag = tag
        self.items = items or []
        self.clickable = clickable

# MetaItem
class MetaItem:
    class Filter:
        def __init__(self, name: str, description: str = None):
            self.name = name
            self.description = description

    def __init__(self, source: object, name: str, icon: object = None, tag: object = None, pakFile: PakFile = None, items: list[MetaItem] = None):
        self.source = source
        self.name = name
        self.icon = icon
        self.tag = tag
        self.pakFile = pakFile
        self.items = items or []

    def findByPath(self, path: str, manager: MetaManager) -> MetaItem:
        paths = re.split('\\\\|/|:', path)
        node = next(iter([x for x in self.items if x.name == paths[0]]), None)
        if node and isinstance(node.source, FileSource) and node.source.pak: node.source.pak.open(node.items, manager)
        # if node and node.pakFile: node.pakFile.open(node.items, manager)
        return node if not node or len(paths) == 1 else node.findByPath(paths[1], manager)

    @staticmethod
    def findByPathForNodes(nodes: list[MetaItem], path: str, manager: MetaManager) -> MetaItem:
        paths = re.split('\\\\|/|:', path, 1)
        node = next(iter([x for x in nodes if x.name == paths[0]]), None)
        if node and isinstance(node.source, FileSource) and node.source.pak: node.source.pak.open(node.items, manager)
        # if node and node.pakFile: node.pakFile.open(node.items, manager)
        return node if not node or len(paths) == 1 else node.findByPath(paths[1], manager)

# IHaveMetaInfo
class IHaveMetaInfo:
    def getInfoNodes(self, resource: MetaManager = None, file: FileSource = None, tag: object = None) -> list[MetaInfo]: pass

# MetaManager
class MetaManager:
    def __init__(self, folderIcon: object = None, packageIcon: object = None):
        self.folderIcon = folderIcon
        self.packageIcon = packageIcon

    def getIcon(self, name: str) -> object: pass

    def getImage(self, name: str) -> object: pass

    @staticmethod
    def _guessStringOrBytes(stream: BytesIO) -> object:
        return stream

    @staticmethod
    def getMetaInfos(manager: MetaManager, pakFile: BinaryPakFile, file: FileSource) -> list[MetaInfo]:
        nodes = None
        obj = pakFile.loadFileObject(object, file)
        match obj:
            case None: return None
            case s if isinstance(obj, IHaveMetaInfo): nodes = s.getInfoNodes(manager, file)
            case s if isinstance(obj, BytesIO):
                value = MetaManager._guessStringOrBytes(s)
                nodes = [
                    MetaInfo(None, MetaContent(type = 'Text', name = 'Text', value = obj)),
                    MetaInfo('Text', items = [
                        MetaInfo(f'Length: {len(obj)}')
                        ])
                ] if isinstance(obj, str) else [
                    MetaInfo(None, MetaContent(type = 'Hex', name = 'Hex', value = obj)),
                    MetaInfo('Bytes', items = [
                        MetaInfo(f'Length: {len(obj)}')
                        ])
                ] if isinstance(obj, BytesIO) else \
                    _throw(f'Unknown {obj}')
        nodes.append(MetaInfo('File', items = [
            MetaInfo(f'Path: {file.path}'),
            MetaInfo(f'FileSize: {file.fileSize}'),
            MetaInfo('Parts', items = [MetaInfo(f'{part.fileSize}@{part.path}') for x in file.parts]) if file.parts else None
            ]))
        # nodes.append(MetaInfo(None, MetaContent(type='Hex',name='TEST',value=BytesIO())))
        return nodes

    @staticmethod
    def getMetaItems(manager: MetaManager, pakFile: BinaryPakFile) -> list[MetaItem]:
        root = []
        if not pakFile.files: return root
        currentPath = None; currentFolder = None

        # parse paths
        for file in sorted(pakFile.files, key = lambda x: x.path):
            # next path, skip empty
            path = file.path[pakFile.pathSkip:]
            if not path: continue
            # folder
            fileFolder = os.path.dirname(path)
            if currentPath != fileFolder:
                currentPath = fileFolder
                currentFolder = root
                if fileFolder:
                    for folder in fileFolder.split('/'):
                        found = next(iter([x for x in currentFolder if x.name == folder and not x.pakFile]), None)
                        if found: currentFolder = found.items
                        else:
                            found = MetaItem(file, folder, manager.folderIcon)
                            currentFolder.append(found)
                            currentFolder = found.items
            # pakfile
            if file.pak:
                items = MetaManager.getMetaItems(manager, file.pak)
                currentFolder.append(MetaItem(file, os.path.basename(file.path), manager.packageIcon, pakFile = pakFile, items = items))
                continue
            # file
            fileName = os.path.basename(path)
            fileNameForIcon = pakFile.fileMask(fileName) or fileName if pakFile.fileMask else fileName
            _, extentionForIcon = os.path.splitext(fileNameForIcon)
            if extentionForIcon: extentionForIcon = extentionForIcon[1:]
            currentFolder.append(MetaItem(file, fileName, manager.getIcon(extentionForIcon), pakFile = pakFile))
        return root